inference-engine/src/mkldnn_plugin/emitters/jit_snippets_emitters.hpp

```diff
@@ -26,39 +46,56 @@ class KernelEmitter : public jit_emitter {
               const std::vector<size_t> &pool = {}, const std::vector<size_t> &gpr = {}) const override {
         emit_impl(in, out, pool, gpr, nullptr);
     }
+    jit_snippets_compile_args _jep;
 
 private:
     void emit_impl(const std::vector<size_t>& in,
                    const std::vector<size_t>& out,
                    const std::vector<size_t>& pool,
                    const std::vector<size_t>& gpr,
                    const MKLDNNPlugin::emitter_context *emit_context) const override {
-        auto tile_rank = in[0]; // count of tile dimensions
-        auto nparams = in[1];
+        const size_t tile_rank = in[0]; // count of tile dimensions
+        const size_t num_inputs = in[1];
+        const size_t num_outputs = in[2];
+        const size_t num_params = num_inputs + num_outputs;
         int reg64_tmp_start { 8 }; // R8, R9, R10, R11, R12, R13, R14, R15 inputs+outputs+1
-        Xbyak::Reg64 params   { dnnl::impl::cpu::x64::abi_param1 };
-        Xbyak::Reg64 schedule { dnnl::impl::cpu::x64::abi_param2 };
-        Xbyak::Reg64 offsets_ { dnnl::impl::cpu::x64::abi_param3 };
-        Xbyak::Reg64 offsets  { dnnl::impl::cpu::x64::abi_not_param1 };
+        Xbyak::Reg64 reg_indexes   { dnnl::impl::cpu::x64::abi_param1 };
+        Xbyak::Reg64 reg_const_params { dnnl::impl::cpu::x64::abi_param2 };
+        #ifdef _WIN32
+        Xbyak::Reg64 reg_tmp_64 { Xbyak::Operand::RDI };
+        #else
+        Xbyak::Reg64 reg_tmp_64 { Xbyak::Operand::RCX };
+        #endif
 
         if (tile_rank != 2)
             IE_THROW() << "Kernel of codegen supports only Tile2D" << std::endl;
 
         h->preamble();
 
-        // to avoid register conflicts on Win systems
-        h->mov(offsets, offsets_);
+        std::vector<Xbyak::Reg64> regs(num_params);
 
-        // ptrs
-        std::vector<Xbyak::Reg64> regs(nparams);
-        for (auto i = 0; i < nparams; i++) {
+        const int64_t harness_num_dims = _jep.output_dims.size() - 1;
```

Comment:
Agree, the check is added

Reply:
There is already a check on the number of parameters in generator.cpp (https://github.com/IvanNovoselov/openvino/blob/9bb9fe46cca907341c9e5f9d3db6bfd823521ffa/src/common/snippets/src/generator.cpp#L56-L58). However, I do agree that the max number of parameters is device-specific, so I moved the check to KernelEmitter.

Reply:
Regarding harness_num_dims, we set an appropriate dims_out to 1 if 2d tile is used (https://github.com/IvanNovoselov/openvino/blob/2494612830208451665e554a31cd348385062fcd/inference-engine/src/mkldnn_plugin/nodes/subgraph.cpp#L445-L448), so it'll be skipped in init_ptrs_with_offsets(). I agree that we can pass an "actual tile rank" in jcp (or derive it from scheduler_dims), and then reduce harness_num_dims if necessary. My concern that it will be even more misleading, because we check that tile_rank must be 2. Probably a better solution is  to avoid adding 2d Tile in the Kenrel if we don't actually use it. @dmitry-gorokhov, what do you think?

Reply:
Sounds reasonable. Lets try to rework it later.